### ISSUE TITLE
Are Model IDs actually stable?

### ISSUE THREAD

**Issue by gajicm93** [airbnb/epoxy#544]
I'm asking about IDs documented here: https://github.com/airbnb/epoxy/wiki/Epoxy-Models#model-ids

The thing is, when an ID is stable, it means that it uniquely represents ALL of the underlying object data, and is not just a "reference to the item" as is the case with classic database index IDs.

For example, if I have a User database table, and a user changes only his name, his index ID stays the same, and that is the example of a NON-STABLE ID. On the other hand, if an ID represents all the backing object data uniquely (as in case of an INT ID returned by hashCode()), that would be a STABLE ID. You can read more about it here: https://stackoverflow.com/a/29446944/4695418

So, what should I use as a Model ID, a classic database index ID that doesn't change when some of the object data items change, or an actual stable ID returned from hashCode?

**Comment by elihart** [comment]
> The thing is, when an ID is stable, it means that it uniquely represents ALL of the underlying object data
This is not true in RecyclerView - in the database world maybe (I'm less familiar with that), but Android defines stable ids differently.

`a classic database index ID that doesn't change when some of the object data items change` this is the right thing to use :)

The stack overflow answer you linked to is wrong. note the comment reply to it
```
-1. This answer is completely wrong. The RecyclerView can only avoid unnecessary work in this way when items have stable ids and you call setHasStableIds(true). The sample code will not work the way the writer says it does under any circumstances. hasStableIds() asks a question about item identities, not about item values. If hasStableIds() returns false, the adapter must recreate all views on every update, as it cannot assume list items with the same ID correspond to the same data. 
```

This is how listview defines stable ids
```
/**
     * Indicates whether the item ids are stable across changes to the
     * underlying data.
     * 
     * @return True if the same id always refers to the same object.
     */
    boolean hasStableIds();
```

`True if the same id always refers to the same object.` - we just need to be able to associate the id to a view, we don't care of the view changes


The point of all this is that this id system lets use detect item insertions, removals, or moves, as well as when an item changes so that we know to rebind its view
